libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -697,11 +697,34 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
 			// Callee removes them and pushes return values
 
 			evmasm::AssemblyItem returnLabel = m_context.pushNewTag();
-			for (unsigned i = 0; i < arguments.size(); ++i)
-				acceptAndConvert(*arguments[i], *function.parameterTypes()[i]);
+
+			if (_functionCall.isSuffixCall() && parameterTypes.size() != 1)
```

Comment:
I'm actually not convinced that we should even implement this for legacy code generation.
The legacy optimizer is not powerful enough to inline and compile-time evaluate even the simplest canonical fixed-point literal cases, so it pretty much never de facto does what people will expect (while via-IR this works well enough).

Reply:
Is that a strong enough reason not to implement it for legacy though? This is also a convenience feature for UDVTs. And in the docs I do not hide the fact that this is just a function call syntax at all so I'm not sure the expectation will really be that strong.

I think it may be better to let people decide for themselves. Or we could issue a warning that would kinda discourage them (and explain the situation) but not hard-disallow this.

Also, having differences between the codegens just increases testing surface and is an unnecessary complication. It's simpler for us when things work the same way.

Reply:
By the way, this may be a good moment to take a look at the docs PR #14121 so that I have a chance to do apply fixes this week, while I'm still available.

Reply:
It's pretty much solely an alternative of having to declare constants with precomputed representations of the UDVT - i.e. ``12.34 f`` is mainly convenience syntax for:
````
Fixed18 constant f12_34 = Fixed18.wrap(12_34 * 10**(18-2)));
````
If the latter is cheaper than the former, people will keep using that, even though it's uglier.

Reply:
I.e. this is *not* an alternative to function calls, but to that really, so it not being more expensive than function calls doesn't really help that much...

Reply:
I still think it's a pretty weak reason for not keeping parity between codegens. In my mind it's a line that I would not cross lightly. But anyway, it's your call.

Reply:
We actually said five features ago that we feature-freeze legacy codegen, I don't know why we still keep implementing stuff for it at all anyways :-).

Reply:
But even if not for that, I'm not sure how "cannot be used for the purpose it is supposed to address" is a weak reason :-D.

Reply:
> We actually said five features ago that we feature-freeze legacy codegen,

Wait, what? If we're in such a freeze now then I don't think it was communicated clearly enough.

I remember some casual discussions that maybe we should stop implementing features for legacy at some point, and I know that we're generally closing issues that do not exist in the new codegen, but I don't think I've heard it explicitly stated that we should no longer be implementing anything for legacy as a general rule. I don't think I've seen us actually doing that in any recent PR and it was not even suggested in the operators PR (which was the most recent major feature for me).

> But even if not for that, I'm not sure how "cannot be used for the purpose it is supposed to address" is a weak reason :-D.

I think it's weak because IMO the response is disproportionate to the problem. I.e. it's not like it being there actively harms anyone - optimizors will ignore it, other people will decide for themselves if they care for gas or for syntax more. And not all uses of the syntax are inefficient anyway. If anything, this encourages people to migrate to IR. Increasing the feature disparity seems like it would be more annoying to users, on the other hand.

But let's talk about it on the call today. This thread is getting long and this topic seems like something that's quicker to discuss in person.